src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs

```diff
@@ -381,5 +430,36 @@ public void BytesRead(int count)
         {
             Interlocked.Add(ref _readTimingBytesRead, count);
         }
+
+        public void StartTimingWrite(int size)
+        {
+            lock (_writeTimingLock)
+            {
+                var minResponseDataRate = _frame.MinResponseDataRate;
+
+                if (minResponseDataRate != null)
+                {
+                    var timeoutTicks = Math.Max(
+                        minResponseDataRate.GracePeriod.Ticks,
+                        TimeSpan.FromSeconds(size / minResponseDataRate.BytesPerSecond).Ticks);
+                    var timeoutTimestamp = (_writeTimingTimeouts.Count == 0
+                            ? _lastTimestamp
+                            : _writeTimingTimeouts.Last()) + timeoutTicks;
```

Comment:
This would be clearer if `timeoutTimestamp  += timeoutTicks` was on its own line. I missed it at first because it looks like it's part of the `false` condition.

Also, I believe the call to `Last()` is going to be expensive. It calls a `Queue.GetEnumerator()` and walks the collection to the end.